### PR TITLE
Avoid double-free in LogCheckError

### PR DESCRIPTION
**Pull request by alexeyr**
The default copy constructor of `LogCheckError` could lead to double-free errors. An alternative to this fix would be to define the constructor and assigment operator (and move constructor?).

### CODE REVIEW

File: include/dmlc/parameter.h [hunk 1]
```diff
@@ -310,7 +310,7 @@ namespace parameter {
 class FieldAccessEntry {
  public:
   FieldAccessEntry()
-      : has_default_(false) {}
+      : has_default_(false), index_(0) {}
```
Comment (hcho3) [inline include/dmlc/parameter.h]: Why are we adding a new field `index_` here?

**Comment by tqchen** [comment]
I believe the reason that a ptr is used is to minimize the cost and allow constant folding. so we'd better keep it as it is

File: include/dmlc/logging.h [hunk 2]
```diff
@@ -66,11 +66,10 @@ inline void InitLogging(const char*) {
...
 class LogCheckError {
  public:
-  LogCheckError() : str(nullptr) {}
-  explicit LogCheckError(const std::string& str_) : str(new std::string(str_)) {}
-  ~LogCheckError() { if (str != nullptr) delete str; }
-  operator bool() {return str != nullptr; }
-  std::string* str;
+  LogCheckError() : str() {}
+  explicit LogCheckError(const std::string& str_) : str(str_) {}
+  operator bool() const { return !str.empty(); }
```
Comment (hcho3) [inline include/dmlc/logging.h]: I think an empty value is different from a missing value. Maybe we should use `std::unique_ptr<std::string>` to store the reference to the temporary string?

Comment (alexeyr) [reply include/dmlc/logging.h]: In general yes, but here it's used for an error message, which should never be empty if present. In particular, it's only called as `LogCheckError("Error.")` or `LogCheckError(os.str())` with non-empty `os`.

File: include/dmlc/parameter.h (hunk 1 above)
Comment (alexeyr) [reply include/dmlc/parameter.h]: We aren't, just initializing it.

**Comment by alexeyr** [comment]
@tqchen The new version keeps the member as pointer.

File: include/dmlc/logging.h (hunk 2 above)
Comment (alexeyr) [reply include/dmlc/logging.h]: Anyway, this is now changed according to @tqchen's comment, probably `unique_ptr` would be as undesirable as my initial solution.

**Pull request merged**